pkg/controller/machineset/controller.go

```diff
@@ -176,57 +176,72 @@ func (r *ReconcileMachineSet) Reconcile(ctx context.Context, request reconcile.R
 	if r.gate.Enabled(featuregate.Feature(openshiftfeatures.FeatureGateMachineAPIMigration)) {
 		machineSetName := machineSet.GetName()
 		machineSetCopy := machineSet.DeepCopy()
-		// Check Status.AuthoritativeAPI. If it's not set to MachineAPI. Set the
-		// paused condition true and return early.
-		//
-		// Once we have a webhook, we want to remove the check that the AuthoritativeAPI
-		// field is populated.
-		if machineSet.Status.AuthoritativeAPI != "" &&
-			machineSet.Status.AuthoritativeAPI != machinev1.MachineAuthorityMachineAPI {
-			conditions.Set(machineSetCopy, conditions.TrueConditionWithReason(
-				machine.PausedCondition,
-				machine.PausedConditionReason,
-				"The AuthoritativeAPI is set to %s", string(machineSet.Status.AuthoritativeAPI),
-			))
-
-			_, err := updateMachineSetStatus(r.Client, machineSet, machineSetCopy.Status)
-			if err != nil && !apierrors.IsNotFound(err) {
-				klog.Errorf("%v: error updating status: %v", machineSetName, err)
-				return reconcile.Result{}, fmt.Errorf("error updating status: %w", err)
-			} else if apierrors.IsNotFound(err) {
-				return reconcile.Result{}, nil
+
+		switch machineSet.Status.AuthoritativeAPI {
+		case "":
+			// An empty .status.authoritativeAPI normally means the resource has not yet been reconciled
+			// by the migration controller, which has the responsibility for propagating .spec.authoritativeAPI to the status.
+			// Pause the resource and take no further action but return until that field is populated.
+			newCondition := conditions.TrueConditionWithReason(
+				machine.PausedCondition, machine.PausedConditionReason,
+				"The AuthoritativeAPI status is not yet set",
+			)
+
+			if !conditions.IsEquivalentTo(conditions.Get(machineSet, machine.PausedCondition), newCondition) {
+				klog.Infof("%v: machine set .status.authoritativeAPI is not yet set, ensuring machine set is paused", machineSetName)
+				conditions.Set(machineSetCopy, newCondition)
+				if _, err := updateMachineSetStatus(r.Client, machineSet, machineSetCopy.Status); err != nil {
+					klog.Errorf("%v: error updating status: %v", machineSetName, err)
+					return reconcile.Result{}, fmt.Errorf("error updating status for machine set %s: %w", machineSetName, err)
+				}
```

Comment:
There is a slight semantic/behavioural change here.
Now in the `.status.AuthoritativeAPI: ""` case we are also setting Paused: True, rather than only just returning, I suppose this is semantically more aligned with our desire of only unpausing the controllers when they are authoritative. @JoelSpeed sounds reasonable?

Reply:
SGTM